src/OpenTelemetry/SimpleExportProcessor.cs

```diff
@@ -21,11 +22,32 @@ public abstract class SimpleExportProcessor<T> : BaseExportProcessor<T>
     protected SimpleExportProcessor(BaseExporter<T> exporter)
         : base(exporter)
     {
+        var exporterType = exporter.GetType();
+        var attributes = exporterType.GetCustomAttributes(typeof(ConcurrencyModesAttribute), true);
+        if (attributes.Length > 0)
+        {
+            var attr = (ConcurrencyModesAttribute)attributes[attributes.Length - 1];
+            this.supportedConcurrencyModes = attr.Supported;
+        }
     }
 
     /// <inheritdoc />
     protected override void OnExport(T data)
     {
+        if (this.supportedConcurrencyModes.HasFlag(ConcurrencyModes.Multithreaded))
```

Comment:
@CodeBlanch do you have some idea how to make this code less ugly? Currently there are lots of duplicated stuff.

Reply:
Most simple thing would probably be like this...

```csharp
public abstract class SimpleExportProcessor<T> : BaseExportProcessor<T>
    where T : class
{
    private readonly object syncObject = new();
    private readonly ConcurrencyModes supportedConcurrencyModes;

    /// <summary>
    /// Initializes a new instance of the <see cref="SimpleExportProcessor{T}"/> class.
    /// </summary>
    /// <param name="exporter">Exporter instance.</param>
    protected SimpleExportProcessor(BaseExporter<T> exporter)
        : base(exporter)
    {
        var exporterType = exporter.GetType();
        var attributes = exporterType.GetCustomAttributes(typeof(ConcurrencyModesAttribute), true);
        if (attributes.Length > 0)
        {
            var attr = (ConcurrencyModesAttribute)attributes[attributes.Length - 1];
            this.supportedConcurrencyModes = attr.Supported;
        }
    }

    /// <inheritdoc />
    protected override void OnExport(T data)
    {
        if (this.supportedConcurrencyModes.HasFlag(ConcurrencyModes.Multithreaded))
        {
            this.Export(data);
            return;
        }

        lock (this.supportedConcurrencyModes.HasFlag(ConcurrencyModes.Global)
            ? GlobalLock.Instance
            : this.syncObject)
        {
            this.Export(data);
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void Export(T data)
    {
        try
        {
            this.exporter.Export(new Batch<T>(data));
        }
        catch (Exception ex)
        {
            OpenTelemetrySdkEventSource.Log.SpanProcessorException(nameof(this.OnExport), ex);
        }
    }
}

internal static class GlobalLock
{
    public static object Instance { get; } = new();
}
```

* Using `MethodImplOptions.AggressiveInlining` so it doesn't introduce an extra call in there.
* Added in the global lock because I was curious what that might look like.

Reply:
Thanks! For "Global", I was thinking of something like named mutex - which works across multi processes.
For example, if OpenTelemetry .NET is used by dotnet-cli (e.g. dotnet.exe) during build, there could be multiple processes running concurrently and outputting metrics/logs/traces to stdout, and we should make sure these are coordinated. This might require more thinking and experiments, most likely I'll scope down this PR by removing "Global" for now.

Reply:
Scoped down this PR.